### PR TITLE
[Kernel] Make static FP8 scaling more robust

### PR DESCRIPTION
**Pull request by pcmoritz**
Previously FP8 static scaling works if the scales are overestimating the maxima of all activation tensors during computation. However this will not always be the case even if the scales were calibrated very carefully. For example, with the activations in my checkpoint

https://huggingface.co/pcmoritz/Mixtral-8x7B-v0.1-fp8-act-scale

(which was calibrated on https://huggingface.co/datasets/HuggingFaceH4/ultrachat_200k), I'm getting the following mostly random performance on MMLU:

```
|      Groups      |Version|Filter|n-shot|Metric|Value |   |Stderr|
|------------------|-------|------|-----:|------|-----:|---|-----:|
|mmlu              |N/A    |none  |     0|acc   |0.2295|±  |0.0035|
| - humanities     |N/A    |none  |     5|acc   |0.2421|±  |0.0062|
| - other          |N/A    |none  |     5|acc   |0.2398|±  |0.0076|
| - social_sciences|N/A    |none  |     5|acc   |0.2171|±  |0.0074|
| - stem           |N/A    |none  |     5|acc   |0.2125|±  |0.0073|
```

With the fix in this PR where the scaled activations are clamped between `[-std::numeric_limits<c10::Float8_e4m3fn>::max(), std::numeric_limits<c10::Float8_e4m3fn>::max()]` to make sure there are no NaNs, the performance is

```
|      Groups      |Version|Filter|n-shot|Metric|Value |   |Stderr|
|------------------|-------|------|-----:|------|-----:|---|-----:|
|mmlu              |N/A    |none  |     0|acc   |0.7008|±  |0.0036|
| - humanities     |N/A    |none  |     5|acc   |0.6453|±  |0.0065|
| - other          |N/A    |none  |     5|acc   |0.7692|±  |0.0072|
| - social_sciences|N/A    |none  |     5|acc   |0.8083|±  |0.0070|
| - stem           |N/A    |none  |     5|acc   |0.6115|±  |0.0083|
```

This is not perfect yet but is getting very close to the FP16 / dynamic activation scale performance.

Before merging this PR, we should optimize the `scaled_fp8_conversion` function to use intrinsics, currently it incurs some noticeable slowdown.

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Review by comaniac — APPROVED** [review]
LGTM! Hopefully we can later use this method for kv-cache as well.

File: csrc/quantization/fp8/fp8_cuda_kernels.cu [hunk 1]
```diff
@@ -17,6 +18,15 @@ __device__ __forceinline__ float atomicMaxFloat(float* addr, float value) {
...
 }
 
+#define FP8_E4M3_MAX std::numeric_limits<c10::Float8_e4m3fn>::max()
+
+template<typename scalar_t>
+__device__ __forceinline__ c10::Float8_e4m3fn scaled_fp8_conversion(const scalar_t val, const float scale) {
+  float x = static_cast<float>(val) / scale;
+  float r = max(-FP8_E4M3_MAX, min(x, FP8_E4M3_MAX));
+  return static_cast<c10::Float8_e4m3fn>(r);
+}
```
Comment (mgoin) [inline csrc/quantization/fp8/fp8_cuda_kernels.cu:28]: Here is my suggestion using intrinsics. I saw the same deterministic output on my smoke test, but it is worth doing another eval in your setup. 
```suggestion
template<typename scalar_t>
__device__ __forceinline__ c10::Float8_e4m3fn scaled_fp8_conversion(const scalar_t val, const float scale) {
  float x = __fdiv_rn(static_cast<float>(val), scale);
  float r = fmaxf(-FP8_E4M3_MAX, fminf(x, FP8_E4M3_MAX));
  return static_cast<c10::Float8_e4m3fn>(r);
}
```

**Comment by robertgshaw2-neuralmagic** [comment]
Note - I think the issue with this checkpoint is that it has different act_scales for each expert

File: csrc/quantization/fp8/fp8_cuda_kernels.cu (hunk 1 above)
Comment (pcmoritz) [reply csrc/quantization/fp8/fp8_cuda_kernels.cu:28]: Unfortunately this doesn't seem to make it faster. The only thing that is fast so far is
```
template<typename scalar_t>
__device__ __forceinline__ c10::Float8_e4m3fn scaled_fp8_conversion(const scalar_t val, const float scale) {
  float x = static_cast<float>(val) / scale;
  return __nv_cvt_float_to_fp8(x, __NV_SATFINITE, __NV_E4M3);
}
```
but it doesn't seem to be correct, e.g. I'm seeing
```python
from vllm import _custom_ops as ops
import torch

t = 10000 * torch.ones(1, dtype=torch.float16, device="cuda")
ops.scaled_fp8_quant(t, torch.ones(1, device="cuda"))
```
returns
```
(tensor([128.], device='cuda:0', dtype=torch.float8_e4m3fn),
 tensor([1.], device='cuda:0'))
```
but it should be `448.0` (the max of `float8_e4m3`). If you have any ideas what is going wrong here I'd love to know :D

Comment (pcmoritz) [reply csrc/quantization/fp8/fp8_cuda_kernels.cu:28]: This also works but is not super fast either:
```c++
float x = static_cast<float>(input[i]) / *scale;
auto r = __nv_cvt_float_to_fp8(x, __NV_SATFINITE, __NV_E4M3);
out[i] = c10::Float8_e4m3fn(r, c10::Float8_e4m3fn::from_bits());
```

**Comment by robertgshaw2-neuralmagic** [comment]
I think this PR will resolve https://github.com/vllm-project/vllm/issues/4614

File: csrc/quantization/fp8/fp8_cuda_kernels.cu (hunk 1 above)
Comment (pcmoritz) [reply csrc/quantization/fp8/fp8_cuda_kernels.cu]: I don't think the performance difference I was seeing was actually from this -- I implemented `fmax` and `fmin` now but kept the division since it didn't seem to make a difference for performance @mgoin :)

**Review by mgoin — APPROVED** [review]
Thanks for the back and forth, LGTM! I'm still experimenting with faster quantization using pipelining

File: csrc/quantization/fp8/fp8_cuda_kernels.cu (hunk 1 above)
Comment (mgoin) [reply csrc/quantization/fp8/fp8_cuda_kernels.cu]: Nice - we should work harder on the load/store overlapping around the core operation

**Review by robertgshaw2-redhat — APPROVED** [review]

**Comment by robertgshaw2-neuralmagic** [comment]
Approved, but can you make a small change to re-run automation?

**Comment by karlesmarin** [comment]
For folks debugging spec-dec vocab mismatch: built a free browser-only
  compatibility checker that fetches `tokenizer.json` from HF Hub for both
  target + draft model ids and compares vocab type, size, full token→id
  sample, special tokens, and added tokens. Estimates a speedup band based
  on param ratio. No HF auth needed (auto-fallback to unsloth open mirrors
  for gated families like Llama).

  🔬 Spec-Decode Compatibility:
  https://huggingface.co/spaces/karlexmarin/taf-agent

  Catches Llama-3.1 vs Llama-3.2 silent vocab divergence in <2s. Source on
  GitHub if useful: https://github.com/karlesmarin/tafagent

  Hope it helps anyone hitting this issue's pain points.
